src/checkers/inference/model/AlwaysFalseConstraint.java

```diff
@@ -0,0 +1,37 @@
+package checkers.inference.model;
+
+import java.util.Collections;
+import org.checkerframework.javacutil.ErrorReporter;
+
+/**
+ * This "constraint" is the result of normalizing another constraint, where that constraint is
+ * always false (evaluates to contradiction). If a constraint is normalized to this constraint, then
+ * an error message will be issued by the {@link ConstraintManager}. This class is implemented as a
+ * singleton.
+ *
+ * @see {@link ConstraintManager}
+ */
+public class AlwaysFalseConstraint extends Constraint {
```

Comment:
We had discussed the use of an enum in the meetings but it was decided that it is better to not use enums, so that there isn't an enum class to maintain on top of creating or deleting the actual constraint class.

T/F "Constraints" are short lived singletons that only exists timeline wise between creation/normalization to `ConstraintManager.add()`. See javadoc of the class.

> Why do we normalize and loose information about a specific kind of Constraint(Subtype, Equality, Inequality) and suddenly go into another subclass of Constraint?

If a particular instance of subtype constraint is always equivalent to equality constraint (eg subtype slot is constant top), there is no information loss by normalizing it here and then allowing for a more efficient encoding in the solver.

> How do you track what the original Constraint was?

Still a design question to be finalized.

Reply:
Update: see `Constraint.normalize()`. The original constraint is kept untouched until just before insertion into the constraint set, where we then normalize it and issue warnings using the original constraint's class name.